src/MSBuild/AutomaticEncodingRestorer.cs

```diff
@@ -0,0 +1,89 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.IO;
+using System.Runtime.InteropServices;
+using System.Security;
+using System.Text;
+
+namespace Microsoft.Build.CommandLine
+{
+    /// <summary>
+    /// Ported from https://github.com/dotnet/sdk/blob/bcea1face15458814b8e53e8785b52ba464f6538/src/Cli/dotnet/AutomaticEncodingRestorer.cs.
+    /// A program can change the encoding of the console which would affect other programs.
+    /// We would prefer to have a pattern where the program does not affect encoding of other programs.
+    /// Create this class in a function akin to Main and let it manage the console encoding resources to return it to the state before execution upon destruction.
+    /// </summary>
+    public class AutomaticEncodingRestorer : IDisposable
+    {
+        private Encoding? _originalOutputEncoding = null;
+        private Encoding? _originalInputEncoding = null;
+
+        private bool outputEncodingAccessible = false;
+        private bool inputEncodingAccessible = false;
+
+        public AutomaticEncodingRestorer()
+        {
+            try
+            {
+                if (
+#if NET7_0_OR_GREATER
+                        !OperatingSystem.IsIOS() && !OperatingSystem.IsAndroid() && !OperatingSystem.IsTvOS()) // Output + Input Encoding are unavailable on these platforms per docs.
+#else
+                        IsWindowsOS()) // Windows is the only platform where we need to change the encoding as other platforms are UTF 8 by default, so for now its the only one required to restore.
+#endif
```

Comment:
If Windows is the only OS where we need it anyway, can we just simplify this to IsWindowsOS()? Or better yet, NativeMethodsShared.IsWindows();

Reply:
That's a good suggestion. For now we only change the encoding on Windows. But IMO the more correct thing to do, which is what the current code does, is to restore the encoding any time it is supported to change it. (In case other encoding changes are made in the future, and nobody reads this code to notice if it only applied to windows.)

https://learn.microsoft.com/en-us/dotnet/api/system.console.inputencoding?view=netframework-4.7.2 It looks like since Android, IOS, Browser, tvOS aren't supported until NET 5 so `InputEncoding` and co are fine to use regardless of OS for Full Framework.

I'll change it to always run on net472. Let me know if you disagree